src/bp/ui-studio/src/web/views/OneFlow/sidePanel/form/SaySomethingForm.tsx

```diff
@@ -0,0 +1,244 @@
+  const formReducer = (state: FormState, action): FormState => {
+    if (action.type === 'resetData') {
+      return {
+        ...state,
+        error: null,
+        contentType: 'builtin_text'
+      }
+    } else if (action.type === 'newData') {
+      const { contentType } = action.data
+
+      return {
+        error: null,
+        contentType: contentType || 'builtin_text'
+      }
+    } else if (action.type === 'updateContentType') {
+      const { value, initial } = action.data
+      const contentType = { contentType: value || 'builtin_text' }
+
+      if (!initial || !state.contentType) {
+        props.updateNode(contentType)
+      }
+
+      return {
+        ...state,
+        ...contentType
+      }
+    } else {
+      throw new Error(`That action type isn't supported.`)
+    }
+  }
+
+  const [formState, dispatchForm] = useReducer(formReducer, defaultFormState)
+  const [showOptions, setShowOptions] = useState(false)
+
+  useEffect(() => {
+    handleContentTypeChange(currentFlowNode?.contentType, true)
+    if (!props.categories?.length) {
+      props.fetchContentCategories()
+    }
+  }, [props.currentFlowNode.id])
+
+  const renameNode = text => {
+    if (text) {
+      const alreadyExists = props.currentFlow.nodes.find(x => x.name === text)
+
+      if (!alreadyExists) {
+        props.updateNode({ name: text })
+      }
+    }
+  }
+
+  const transformText = text => {
+    return text.replace(/[^a-z0-9-_\.]/gi, '_')
+  }
+
+  const onCopy = () => {
+    props.copyFlowNode()
+    setShowOptions(false)
+    toastInfo('Copied to buffer')
+  }
+
+  const handleContentTypeChange = (value, initial = false) => {
+    dispatchForm({ type: 'updateContentType', data: { value, initial } })
+  }
+
+  const { currentFlowNode, readOnly } = props
+  const { contentType } = formState
+  const categories = props.categories?.filter(cat => shownCategories.includes(cat.id))
+
+  const moreOptionsItems: MoreOptionsItems[] = [
+    {
+      icon: 'duplicate',
+      label: 'Copy',
+      action: onCopy.bind(this)
+    },
+    {
+      icon: 'trash',
+      label: 'Delete',
+      action: props?.onDeleteSelectedElements,
+      className: MoreOptionsStyles.delete
+    }
+  ]
+
+  const goThroughObjectAndLeaveOutIndex = (properties, indexToRemove) => {
+    const returnObject = {}
+
+    Object.keys(properties).forEach(key => {
+      if (key !== indexToRemove) {
+        returnObject[key] =
+          Object.prototype.toString.call(properties[key]) === '[object Object]'
+            ? goThroughObjectAndLeaveOutIndex(properties[key], indexToRemove)
+            : properties[key]
+      }
+    })
+
+    return returnObject
+  }
+
+  const removeDescriptions = json => {
+    const { properties, ...leftover } = json
+
+    const newProperties = goThroughObjectAndLeaveOutIndex(properties, 'description')
+
+    return { properties: newProperties, ...leftover }
+  }
+
+  const getCurrentCategory = () => {
+    if (!contentType || !categories) {
```

Comment:
These should definitely be renamed, it's contentTypes, not categories. I think we could safely rename `state.content.categories` to `state.content.contentTypes` and stick with contentType everywhere